Add MongoDB filtering and deletion, and various upgrades

katcharov opened the pull request:
<!--
Thank you so much for your contribution!

Please fill in all the sections below.
Please open the PR as a draft initially. Once it is reviewed and approved, we will ask you to add documentation and examples.
Please note that PRs with breaking changes or without tests will be rejected.

Please note that PRs will be reviewed based on the priority of the issues they address.
We ask for your patience. We are doing our best to review your PR as quickly as possible.
Please refrain from pinging and asking when it will be reviewed. Thank you for understanding!
-->

## Issue
<!-- Please specify the ID of the issue this PR is addressing. For example: "Closes #1234" or "Fixes #1234" -->
Closes #1868
Closes #1532

## Change
<!-- Please describe the changes you made. -->

* Implement mongodb filtering and deletion (#1868)
* Move test boilerplate into a TestHelper
* Throw an exception if no index exists once the store is created (addresses #1532)
* Upgrade mongodb-driver-sync to 5.2.0
* Upgrade from legacy knnVector to vectorSearch
* Minor refactoring and fixes

## General checklist
<!-- Please double-check the following points and mark them like this: [X] -->
- [X] There are no breaking changes
- [X] I have added unit and integration tests for my change
- [X] I have manually run all the unit and integration tests in the module I have added/changed, and they are all green
- [ ] **No changes to core.** I have manually run all the unit and integration tests in the [core](https://github.com/langchain4j/langchain4j/tree/main/langchain4j-core) and [main](https://github.com/langchain4j/langchain4j/tree/main/langchain4j) modules, and they are all green
<!-- Before adding documentation and example(s) (below), please wait until the PR is reviewed and approved. -->
- [ ] I have added/updated the [documentation](https://github.com/langchain4j/langchain4j/tree/main/docs/docs)
- [ ] I have added an example in the [examples repo](https://github.com/langchain4j/langchain4j-examples) (only for "big" features)
- [ ] I have added/updated [Spring Boot starter(s)](https://github.com/langchain4j/langchain4j-spring) (if applicable)

## Checklist for changing existing embedding store integration
<!-- Please double-check the following points and mark them like this: [X] -->
- [ ] I have manually verified that the `{NameOfIntegration}EmbeddingStore` works correctly with the data persisted using the latest released version of LangChain4j

katcharov committed:
Add MongoDB filtering and deletion, and various upgrades

* Implement mongodb filtering and deletion
* Move test boilerplate into a TestHelper
* Throw an exception if no index exists once the store is created
* Upgrade mongodb-driver-sync to 5.2.0
* Upgrade from legacy knnVector to vectorSearch
* Minor refactoring and fixes

katcharov committed:
Refactor to use MongoDbTestFixture

katcharov commented on langchain4j-core/src/test/java/dev/langchain4j/store/embedding/EmbeddingStoreWithoutMetadataIT.java:
```diff
@@ -356,7 +356,13 @@ void should_return_correct_score() {
                 .build()).matches()).isEqualTo(relevant);
     }
 
+    @Deprecated
```
I extracted this to a static method, to use in a test class that did not extend this class. It seems like a generally useful utility method.

I added this to discourage use of this variant, since it seemed only necessary for a workaround. Looking closer, when Awaitility was added, it did not replace the sleep, but was [added before the sleep](https://github.com/langchain4j/langchain4j/pull/1610/files#diff-4c1e931865cee83990f37441139cbc2b0052a06b818f9d0396a9909f7a702dd7R326). I believe the sleep is no longer needed, but I am not set up to test Azure. I could add the following [commit](https://github.com/langchain4j/langchain4j/commit/99eddce0dcaa9c894acfa6490461bc1dcaf48dee) (removal of extra sleep) to this PR, or I could get rid of the Deprecated, or I could leave it. Let me know what you think.

katcharov replied:
Applied

katcharov commented:
As part of rebasing on #1859, I enabled container tests and [added a check](https://github.com/langchain4j/langchain4j/pull/1869/files#diff-cf0c855935403e01addf947679edb71504506055aa795c1398c3219133232373R416) for PENDING state to fix failures.

katcharov committed:
Rename to MongoDbEmbeddingStoreIT

Martin7-1 reviewed (changes requested):
Looks like you change from `Atlas Search` to `Atlas Vector Search` right? Could you please check the javadoc in `MongoDbEmbeddingStore`?

Also, it's a breaking change, maybe we need documents to tell users how to migrate it from `AtlasSearch` to `AtlasVectorSearch` in detail.

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/TestHelper.java:
```diff
@@ -0,0 +1,116 @@
+        Assumptions.assumeTrue(hasParts || CONNECTION_STRING != null,
+                "Skipping test: host/username/password, or connection string must be provided via env vars.");
+
+        String connectionString = CONNECTION_STRING != null
```
Looks like `CONNECTION_STRING` is always non-null. How to test `MongoDB Atlas` on the cloud?

*additions:*

After double check your code, maybe we could make `CONNECTION_STRING` a environment variable so that we don't need to change the code.

katcharov replied:
You are right, this should be an env var. I unintentionally introduced the hardcoded value during a rebase, thank you.

Just to explain why the connection string is here at all, I used it as another option for testing with [local Atlas CLI](https://www.mongodb.com/docs/atlas/cli/current/atlas-cli-deploy-local/) rather than cloud (the recently-updated container seems just as good for local testing).

katcharov replied:
I added this as an env var, resolved.

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/TestHelper.java:
```diff
@@ -0,0 +1,116 @@
+    }
+
+    void afterTests() {
+        getDatabase().drop();
```
Looks like `MongoDB Atlas Cloud` does not allow user to drop database. (I don't find a way to trigger `MongoDB Atlas Cloud` test, so I change `CONNECTION_STRING` to `null` manually) 

```java
com.mongodb.MongoCommandException: Command failed with error 8000 (AtlasError): 'user is not allowed to do action [dropDatabase] on [test_database_1062281260.]' on server ac-fte18bh-shard-00-02.eo9fifu.mongodb.net:27017. The full response is {"ok": 0, "errmsg": "user is not allowed to do action [dropDatabase] on [test_database_1062281260.]", "code": 8000, "codeName": "AtlasError"}

	at com.mongodb.internal.connection.ProtocolHelper.getCommandFailureException(ProtocolHelper.java:210)
	at com.mongodb.internal.connection.InternalStreamConnection.receiveCommandMessageResponse(InternalStreamConnection.java:520)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceiveInternal(InternalStreamConnection.java:448)
	at com.mongodb.internal.connection.InternalStreamConnection.lambda$sendAndReceive$0(InternalStreamConnection.java:375)
	at com.mongodb.internal.connection.InternalStreamConnection.sendAndReceive(InternalStreamConnection.java:378)
	at com.mongodb.internal.connection.UsageTrackingInternalConnection.sendAndReceive(UsageTrackingInternalConnection.java:111)
	at com.mongodb.internal.connection.DefaultConnectionPool$PooledConnection.sendAndReceive(DefaultConnectionPool.java:748)
	at com.mongodb.internal.connection.CommandProtocolImpl.execute(CommandProtocolImpl.java:68)
	at com.mongodb.internal.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:208)
	at com.mongodb.internal.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:113)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:82)
	at com.mongodb.internal.connection.DefaultServerConnection.command(DefaultServerConnection.java:73)
	at com.mongodb.internal.connection.DefaultServer$OperationCountTrackingConnection.command(DefaultServer.java:298)
	at com.mongodb.internal.operation.SyncOperationHelper.executeCommand(SyncOperationHelper.java:231)
	at com.mongodb.internal.operation.DropDatabaseOperation.lambda$execute$0(DropDatabaseOperation.java:61)
	at com.mongodb.internal.operation.SyncOperationHelper.withConnectionSource(SyncOperationHelper.java:167)
	at com.mongodb.internal.operation.SyncOperationHelper.withConnection(SyncOperationHelper.java:112)
	at com.mongodb.internal.operation.DropDatabaseOperation.execute(DropDatabaseOperation.java:60)
	at com.mongodb.internal.operation.DropDatabaseOperation.execute(DropDatabaseOperation.java:45)
	at com.mongodb.client.internal.MongoClusterImpl$OperationExecutorImpl.execute(MongoClusterImpl.java:379)
	at com.mongodb.client.internal.MongoDatabaseImpl.executeDrop(MongoDatabaseImpl.java:230)
	at com.mongodb.client.internal.MongoDatabaseImpl.drop(MongoDatabaseImpl.java:220)
	at dev.langchain4j.store.embedding.mongodb.TestHelper.afterTests(TestHelper.java:101)
	at dev.langchain4j.store.embedding.mongodb.MongoDbEmbeddingStoreLegacyIndexesIT.afterEach(MongoDbEmbeddingStoreLegacyIndexesIT.java:50)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
```

Maybe we need the original way to reuse the same database and delete all data in it like below:

```java
    @Override
    protected void clearStore() {
        CodecRegistry pojoCodecRegistry = fromProviders(PojoCodecProvider.builder()
                .register(MongoDbDocument.class, MongoDbMatchedDocument.class)
                .build());
        CodecRegistry codecRegistry = fromRegistries(MongoClientSettings.getDefaultCodecRegistry(), pojoCodecRegistry);

        MongoCollection<MongoDbDocument> collection = client.getDatabase(DATABASE_NAME)
                .getCollection(COLLECTION_NAME, MongoDbDocument.class)
                .withCodecRegistry(codecRegistry);

        Bson filter = Filters.exists("embedding");
        collection.deleteMany(filter);
    }
```

Also, `MongoDB Atlas Cloud` only allow create 3 indexes on M0 clusters. So maybe reuse the same database and collection is a better way to test. WDYT?

katcharov replied:
There are a couple of related issues here: whether the databases should be dropped during tests, and the implications for use of M0 Atlas for testing.

I used a new database with a random name for each test in this PR. The indexes are eventually consistent, so it is possible for tests to fail in confusing ways. This seemed easier to debug and less error-prone, and it was not slower. I believe that if it is possible to test in this way, it is preferable.

It should be possible to test against only the `MongoDBAtlasLocalContainer`. This (along with the local Atlas CLI) are the recommended ways to test. I did not want to remove the M0 Atlas tests in this PR without discussing first, but I propose testing exclusively with the container. Though, if there is some reason not to do this, that is fine, and I would look into how to address that. Let me know!

Martin7-1 replied:
1. For local container, it doesn't matter right? For cloud, we reuse the same database and collection at any test time, and we will delete all datas before each test. And just keep the database and collection on cloud.
2. Tests in `Atlas Cloud` is just for consistence. We don't want to see that users failed to embed in the cloud. (Because we don't know the difference between cloud and local, such as cloud do not allow users to drop database using command).

katcharov replied:
1. The search indexes are still eventually consistent on a local container. The index itself is technically data - to give an example similar to something I ran into during this PR, consider an index that has been marked for deletion still being present when the next test runs and attempts to use/create that index. Dropping will ensure that each test starts with a clean slate.

2. I believe there is no difference between cloud and local. The drop database problem above appears to be a [permissions issue](https://stackoverflow.com/questions/55636442), and the index creation limitation was removed in August (this [page](https://www.mongodb.com/docs/atlas/atlas-search/shared-tier-limitations/) discusses limitations, and will be updated soon to reflect this).

If there is no longer an objection to removing cloud tests (2), then I think this would resolve the problem. Otherwise, perhaps we could try updating the permissions for the test account?

Martin7-1 replied:
Sounds reasonable, maybe we can discard Cloud test after `MongoDB Atlas` update the drop database permission. And now could we just keep it?

katcharov replied:
Apologies, I should clarify - the drop database permission is something that users who plan to test against cloud would need to have configured on their Atlas account. The default account has this permission, but if you create secondary accounts, those will not by default. However, to avoid this matter entirely, I have changed the tests to use a random collection name (instead of a database) in this [PR](https://github.com/langchain4j/langchain4j/pull/1932), and rebased. I believe this removes any impediment to removing the cloud test?

I have set tests up to check if connection string or host/user/pass env vars are set. If they are not set, the test will use local container testing. The env vars are present now in case devs want to use Atlas CLI, and save a few seconds on test startup. I have removed the env vars from the workflow, and this should disable cloud testing. I have also checked with other developers and there should be no issue with testing against the container, especially after the recent upgrade to the MongoDB testcontainer.

I ran the PR's tests against a) the local container, and b) a cloud instance using an account that had write but not admin permissions, and they passed.

katcharov replied:
I believe this is resolved, we agreed that testing against the testcontainer should be fine. I confirmed internally with others at MongoDB that this is the recommended approach. I have adjusted it to use random collections rather than databases, so it still works against M0, if someone were to configure it this way locally or in the future.

Martin7-1 reviewed (changes requested):
@katcharov Thank you!

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,57 @@
+        }
+    }
+
+    private TestHelper helper = new TestHelper(createClient()).initialize();
```
I think we should make all `TestHelper` static, in order to reuse the database and collection.

katcharov replied:
Replied in the [above comment](https://github.com/langchain4j/langchain4j/pull/1869#discussion_r1796161921).

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,57 @@
+
+public class MongoDbEmbeddingStoreIT extends EmbeddingStoreWithFilteringIT {
+
+    public static class ContainerIT extends MongoDbEmbeddingStoreWithRemovalIT {
```
Would it be better to move this class to a separate file? I'm not sure if integration-test will be able to execute the class this way.

BTW, should `ContainerIT` extend `MongoDbEmbeddingStoreIT` rather than `MongoDbEmbeddingStoreWithRemovalIT`?

katcharov replied:
Good catch on the extends, thanks!

I do not mind moving these to new files, will have a look.

Martin7-1 replied:
On the other hand, if we keep Cloud test and Local test in the same file, Could you please add `@EnabledIfEnvironmentVariable` in cloud tests to see whether `ContainerIT` will be triggered normally?

katcharov replied:
Resolved, we are using just the testcontainer now, and this is done by checking the env vars in an if statement (since we can now also specify just a connection string).

katcharov committed:
Update docs, fixes, renames

katcharov commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreLegacyIndexesIT.java:
```diff
@@ -0,0 +1,74 @@
+        helper.getDatabase().createCollection("test_collection");
+        MongoCollection<Document> collection = helper.getDatabase().getCollection("test_collection");
+        collection.createSearchIndex("test_index_legacy", fromIndexMappingLegacyKnnVector(indexMapping));
+        return helper.initialize(builder -> builder.createIndex(false).indexName("test_index_legacy"));
```
@Martin7-1 :
> Also, it's a breaking change, maybe we need documents to tell users how to migrate it from AtlasSearch to AtlasVectorSearch in detail.

I believe this is not a breaking change. The tests here create the old index (I have appended `_legacy` for clarity, though I may need to revert this for M0 tier testing), but they still pass when using the updated aggregation stage.

I agree the docs should include information about upgrading, though this might be as simple as telling users to delete their old index (and re-create manually if needed).

Martin7-1 replied:
OK. got it.

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreIT.java:
```diff
@@ -0,0 +1,57 @@
+
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.*;
+
+public class MongoDbEmbeddingStoreIT extends EmbeddingStoreWithFilteringIT {
```
Minor: remove public to keep consistence

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreLegacyIndexesIT.java:
```diff
@@ -0,0 +1,74 @@
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.EMBEDDING_MODEL;
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.createClientFromEnv;
+
+public class MongoDbEmbeddingStoreLegacyIndexesIT extends EmbeddingStoreIT {
```
Minor: remove public to keep consistence

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreLegacyIndexesIT.java:
```diff
@@ -0,0 +1,74 @@
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.EMBEDDING_MODEL;
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.createClientFromEnv;
+
+public class MongoDbEmbeddingStoreLegacyIndexesIT extends EmbeddingStoreIT {
```
I'm curious about why this class has no `ContainerIT`?

katcharov replied:
(Resolved, all classes now use the container through the fixture)

Martin7-1 commented on langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreWithRemovalIT.java:
```diff
@@ -0,0 +1,57 @@
+
+import static dev.langchain4j.store.embedding.mongodb.MongoDbTestFixture.*;
+
+public class MongoDbEmbeddingStoreWithRemovalIT extends EmbeddingStoreWithRemovalIT {
```
Minor: remove public to keep consistence

katcharov commented:
I have created the following, which split out refactoring as requested:
* https://github.com/langchain4j/langchain4j/pull/1931
* https://github.com/langchain4j/langchain4j/pull/1932

katcharov committed:
Use lazy singleton testcontainer, rename, use awaitUntilAsserted

vbabanin commented on langchain4j-core/src/test/java/dev/langchain4j/store/embedding/TestUtils.java:
```diff
@@ -0,0 +1,19 @@
+
+import java.time.Duration;
+
+public class TestUtils {
```
```suggestion
public final class TestUtils {
```

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -268,40 +276,66 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
-    public List<EmbeddingMatch<TextSegment>> findRelevant(Embedding referenceEmbedding, int maxResults, double minScore) {
-        List<Double> queryVector = referenceEmbedding.vectorAsList().stream()
+    public void removeAll() {
+        collection.deleteMany(new Document());
```
[nit] Just to be more explicit. 
```suggestion
        collection.deleteMany(Filters.empty());
```

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -321,14 +355,17 @@ private void addAllInternal(List<String> ids, List<Embedding> embeddings, List<T
+        if (!result.wasAcknowledged()) {
             String errMsg = String.format("[MongoDbEmbeddingStore] Add document failed, Document=%s", documents);
-            log.warn(errMsg);
+            if (log.isErrorEnabled()) {
```
SLF4J handles this internally and won’t log unless the level is set to `ERROR`, so it seems we could safely remove this if statement.

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -268,40 +276,66 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
         } catch (MongoCommandException e) {
             if (log.isErrorEnabled()) {
-                log.error("Error in MongoDBEmbeddingStore.findRelevant", e);
+                log.error("Error in MongoDBEmbeddingStore.search", e);
```
SLF4J handles this check `log.isErrorEnabled()` internally and won’t log unless the level is set to `ERROR`, so it seems we could safely remove the if statement above.

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -268,40 +276,66 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
-                ));
+        long numCandidates = request.maxResults() * maxResultRatio;
+
+        Bson filter = null;
```
Should we give this filter and the one defined as the field more descriptive names? Currently, they have the same name, but one functions as a `global pre-filter` (could be named globalPreFilter) and the other as a `current request post-filter`(could be named currentRequestPostFilter), which could cause confusion.

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -342,13 +379,43 @@ private void createCollection(MongoDatabase database, String collectionName, Cre
+        long startTime = System.nanoTime();
+        long timeoutNanos = TimeUnit.SECONDS.toNanos(SECONDS_TO_WAIT_FOR_INDEX);
+        while (System.nanoTime() - startTime < timeoutNanos) {
+            if (indexExistsAndNot(indexName, "INITIAL_SYNC", "PENDING", "BUILDING")) {
```
Index also might end up being in `FAILED` status which they can enter due to an invalid index definition. Should we throw an exception in this case if we have detected a failure within a given waiting time period?

katcharov replied:
I have added this, and also used the `queryable` flag to check if we should proceed (instead of the list of statuses).

katcharov committed:
Use random collection name instead of random db name

vbabanin commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -268,40 +276,66 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
+                metaVectorSearchScore("score"),
+                include("embedding", "metadata", "text"))));
+        if (filter != null) {
+            Bson match = match(filter);
```
[Optional] The request filter currently always applies in post-filtering. If the request doesn’t involve score filtering and all request filter fields are indexed in the index mapping, could we potentially move the filter to the pre-filter stage? We could potentially check this dynamically using `IndexMapping.` However, I'm not sure how common queries without score filtering are.

katcharov replied:
`minScore` should always be present, since it is a double, with null defaulting to 0.

Earlier, I attempted to move most of the filtering into the prefilter (this is the suggested approach), however, I ran into issues and gave this up.

katcharov committed:
Remove non-container testing, minor fixes, refactoring and cleanup

caseyclements commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -268,41 +277,65 @@ public List<String> addAll(List<Embedding> embeddings, List<TextSegment> embedde
+                vectorSearchOptions));
+        pipeline.add(project(fields(
+                metaVectorSearchScore("score"),
+                include("embedding", "metadata", "text"))));
```
I've been advised to default to avoid including embedding vectors in the output.

katcharov replied:
Looks like it's needed for the tests to pass. But perhaps there should be an option either in the Mongo builder, or at the general EmbeddingSearchRequest level, to omit this. Maybe that's more complex than fits in with the approach taken here, I don't know. I'll leave this up to the maintainers.

katcharov replied:
Resolving as out of scope for PR (though worthwhile to consider).

caseyclements commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStore.java:
```diff
@@ -342,13 +376,48 @@ private void createCollection(MongoDatabase database, String collectionName, Cre
+            }
+        }
+        log.warn("Index {} was not created or did not exit INITIAL_SYNC within {} seconds",
+                indexName, SECONDS_TO_WAIT_FOR_INDEX);
```
We wait on this condition. 
`index["type"] == "vectorSearch" and index["status"] == "READY"`

Determining when all of your new documents have been indexed is a more difficult challenge. The only thing I've found that works is to loop over vector searches and make sure than I get back the number that I requested, what you call cleverly, maxRequests.

katcharov replied:
Still looking into this. If there is a difference, it seems like it will be subtle.

katcharov replied:
This is fine. The only relevant difference is that a failed index might be queryable, but we check for failure directly above, so there is no difference here. Resolving.

caseyclements commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbMatchedDocument.java:
```diff
@@ -9,7 +9,7 @@ public class MongoDbMatchedDocument {
     private List<Float> embedding;
     private String text;
-    private Map<String, String> metadata;
+    private Map<String, Object> metadata;
```
We wrote a utility to convert ObjectIds to strings within metadata, as there was some requirement that they wished metadata to be serialized to json. Just something to keep in mind for future.

katcharov replied:
Could you link, just for reference?

caseyclements commented on langchain4j-mongodb-atlas/src/main/java/dev/langchain4j/store/embedding/mongodb/MongoDbMetadataFilterMapper.java:
```diff
@@ -0,0 +1,95 @@
+        } else {
+            throw new UnsupportedOperationException("Unsupported filter type: " + filter.getClass().getName());
+        }
+    }
```
This is great. We do all our MQL as dictionaries with a ton of strings.

katcharov replied:
I believe this is referring to the Python MongoDB driver, which @caseyclements is a developer on.

caseyclements reviewed (approved):
LGTM. 
Comments for discussion and sharing experience.

Martin7-1 reviewed (approved):
LGTM. Cloud you please check docs in `/docs` if there need any updates?

katcharov committed:
Update docs

katcharov commented:
I have updated the docs, looking at only the two files in that last commit (not sure if there are docs elsewhere). They're a bit sparse, but I added a note about Atlas CLI, and the fact that there are plans to release into Community Edition (so, not just Atlas). I removed the examples section linking to the tests - after this PR, the tests that were linked there are even less ideal, as examples. I opened up a ticket on our end to add an example to the examples repo instead, though perhaps someone else will get to it first.

Note that review comments are outstanding on https://github.com/langchain4j/langchain4j/pull/1932 , which this PR is based on (I believe I addressed what seemed possible to, though).

katcharov committed:
Remove debug code

vbabanin reviewed (approved):
LGTM!

dliubarskyi committed:
Merge branch 'main' into java-improvements

dliubarskyi committed:
Merge branch 'main' into java-improvements-refactor2

dliubarskyi committed:
Merge branch 'main' into java-improvements-refactor2

katcharov committed:
Merge branch 'main' into java-improvements-refactor2

katcharov committed:
Apply suggestions from code review

Co-authored-by: LangChain4j <langchain4j@gmail.com>

katcharov committed:
Link to updated test under example, fix import

katcharov commented on docs/docs/integrations/embedding-stores/mongodb-atlas.md:
```diff
@@ -31,9 +34,3 @@ To use Atlas Vector search, you need to have an Atlas deployment. MongoDB offers
 
 - `MongoDbEmbeddingStore`
-
-## Examples
```
I restored a link to the class (the cloud vs local options are now handled in in the same way, in the fixture).

The test files don't have much code, but I suppose they could use them to see how the fixture is used and go from there.

katcharov replied:
I think it should work when this PR is merged, I had renamed the file.

katcharov commented:
@langchain4j 

> The tests are failing for me locally, could you please check?

I was able to partially reproduce the error, but this was by shutting down the server that I was trying to connect to, and the extended message was slightly different.

Otherwise, tests pass for me. I ran the tests directly from IntelliJ. I did not have any of the env vars set (`MONGODB_CONNECTION_STRING`, `MONGODB_ATLAS_HOST`, ...), and it defaulted to using the testcontainer. Do you have any of the env vars set, so as to test against some already-running instance? If you clear any such env vars, does it fall back to testing against the testcontainer? (This needs docker, as it did before this PR.)

Even if the env vars are set, it should still work. When I tested, I successfully ran tests against both a local instance, and an Atlas M0 instance, via a connection string. From the error, it looks like it cannot connect at `ac-gpxo1fa-shard-00-00.1x5hb8d.mongodb.net:27017`. Perhaps confirm that the env vars are correct? Maybe something has changed in the environment - are the results the same when running the tests on the main branch?

katcharov committed:
Merge branch 'main' into java-improvements-refactor2

katcharov committed:
PR cleanup

katcharov committed:
Merge branch 'java-improvements-refactor2' into java-improvements

# Conflicts:
#	langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreIT.java
#	langchain4j-mongodb-atlas/src/test/java/dev/langchain4j/store/embedding/mongodb/MongoDbEmbeddingStoreMiscIT.java

katcharov committed:
Remove MongoDB env vars

katcharov commented:
@langchain4j I have merged in changes from #1932, and I have also removed env vars from CI, as I did in https://github.com/langchain4j/langchain4j/pull/1869/commits/1eadde9a8b5eaa331155f8e4f7c2ab9b8f88195d (I am not sure if I missed these, or if a merge added them).

I have also renamed the env vars to remove "Atlas" (note that vector search is not planned to be limited to Atlas, and is in [early access](https://www.mongodb.com/resources/beta-programs#early-access) for Community Edition).

I think this might ensure that the tests are running on the container for you locally.

dliubarskyi committed:
Merge branch 'main' into java-improvements

katcharov commented:
@langchain4j Thanks! I have gone through the comments, and resolved them, leaving a comment explaining the resolution where needed.

katcharov commented:
@langchain4j no worries, and thank you!
